### PR TITLE
Implement consensus state pruning for Tendermint clients

### LINKED ISSUE #600: Tendermint light client: add consensus state pruning on client update
ibc-go's tendermint light client [prunes](https://github.com/cosmos/ibc-go/blob/ae27fa5960d523e4476776308ecd0b4cd3377671/modules/light-clients/07-tendermint/update.go#L137) old consensus states that can no longer be used  upon updating the client with a new consensus states. This is a great idea as it keeps the storage requirements of the light client constant over time.

Pruning on every update might be needlessly expensive though; we could even make the point where pruning runs configurable (e.g. every X days, or every Y block, etc).

### PR DESCRIPTION
**Pull request by seanchen1991**
Closes: #600 

## Description

Implements consensus state pruning for Tendermint clients. 

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/src/core/ics02_client/context.rs [hunk 1]
```diff
@@ -53,6 +53,30 @@ pub trait ClientExecutionContext: Sized {
...
+        height: Height,
+    ) -> Result<(), ContextError>;
+
+    /// Delete the update height associated with the client at the specified height. This update
+    /// time should be associated with a consensus state through the specified height.
+    fn delete_update_height(
+        &mut self,
+        client_id: ClientId,
+        height: Height,
+    ) -> Result<(), ContextError>;
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/ics02_client/context.rs]: Would you place these two after the `store_update_*` methods?

File: crates/ibc/src/clients/ics07_tendermint/client_state/update_client.rs [hunk 2]
```diff
@@ -163,4 +165,47 @@ impl ClientState {
...
+                CommonContext::consensus_state(ctx, &client_consensus_state_path)?;
+            let tm_consensus_state: TmConsensusState =
+                consensus_state
+                    .try_into()
+                    .map_err(|err| ClientError::Other {
+                        description: err.to_string(),
+                    })?;
+
+            let host_timestamp = ctx.host_timestamp()?;
+            let tm_consensus_state_timestamp = (tm_consensus_state.timestamp() + self.trusting_period)
```
Comment (Farhad-Shabani) [inline crates/ibc/src/clients/ics07_tendermint/client_state/update_client.rs]: Can we rename `tm_consensus_state_timestamp` to something like `tm_consensus_state_expiry` for more clarity?

File: crates/ibc/src/clients/ics07_tendermint/client_state.rs [hunk 3]
```diff
@@ -488,7 +490,7 @@ where
 
 impl<E> ClientStateExecution<E> for ClientState
 where
-    E: TmExecutionContext,
+    E: TmExecutionContext + ExecutionContext + TmValidationContext,
```
Comment (Farhad-Shabani) [inline crates/ibc/src/clients/ics07_tendermint/client_state.rs]:
```suggestion
    E: TmExecutionContext + ExecutionContext,
```

File: crates/ibc/src/clients/ics07_tendermint/client_state/update_client.rs [hunk 4]
```diff
@@ -163,4 +165,47 @@ impl ClientState {
...
         }
     }
+
+    pub fn prune_oldest_consensus_state<E>(
+        &self,
+        ctx: &mut E,
+        client_id: &ClientId,
+    ) -> Result<(), ClientError>
+    where
+        E: ClientExecutionContext + TmValidationContext,
```
Comment (Farhad-Shabani) [inline crates/ibc/src/clients/ics07_tendermint/client_state/update_client.rs]:
```suggestion
        E: ClientExecutionContext + CommonContext,
```

Comment (seanchen1991) [reply crates/ibc/src/clients/ics07_tendermint/client_state/update_client.rs]: [1bd8588](https://github.com/cosmos/ibc-rs/pull/922/commits/1bd85885bb04d0a495af414ebe755487ed0d5f7a)

File: crates/ibc/src/clients/ics07_tendermint/client_state.rs (hunk 3 above)
Comment (seanchen1991) [reply crates/ibc/src/clients/ics07_tendermint/client_state.rs]: [1bd8588](https://github.com/cosmos/ibc-rs/pull/922/commits/1bd85885bb04d0a495af414ebe755487ed0d5f7a)

File: crates/ibc/src/clients/ics07_tendermint/client_state/update_client.rs (hunk 2 above)
Comment (seanchen1991) [reply crates/ibc/src/clients/ics07_tendermint/client_state/update_client.rs]: [1bd8588](https://github.com/cosmos/ibc-rs/pull/922/commits/1bd85885bb04d0a495af414ebe755487ed0d5f7a)

File: crates/ibc/src/core/ics02_client/context.rs (hunk 1 above)
Comment (seanchen1991) [reply crates/ibc/src/core/ics02_client/context.rs]: [1bd8588](https://github.com/cosmos/ibc-rs/pull/922/commits/1bd85885bb04d0a495af414ebe755487ed0d5f7a)

**Review by Farhad-Shabani — APPROVED** [review]
All set from my end. See if the added unit test makes sense to you, we can merge. 🎉 
Thanks @seanchen1991 for taking care of this!

**Pull request merged**